packages/flutter/lib/src/material/ink_sparkle.dart

```diff
@@ -0,0 +1,3647 @@
+       _clipCallback = _getClipCallback(referenceBox, containedInkWell, rectCallback),
+       super(controller: controller, referenceBox: referenceBox, color: color, onRemoved: onRemoved) {
+    // InkSparkle will not be painted until the async compilation completes.
+    _InkSparkleFactory.compileShaderIfNeccessary();
+    
+    controller.addInkFeature(this);
+
+    // Immediately begin animating the ink.
+
+    _animationController = AnimationController(
+      duration: _animationDuration,
+      vsync: controller.vsync,
+    )..addListener(controller.markNeedsPaint)..forward();
+
+    _radiusScale = TweenSequence<double>(
+      <TweenSequenceItem<double>>[
+        TweenSequenceItem<double>(
+          tween: CurveTween(curve: Curves.fastOutSlowIn),
+          weight: 75,
+        ),
+        TweenSequenceItem<double>(
+          tween: ConstantTween<double>(1.0),
+          weight: 25,
+        ),
+      ],
+    ).animate(_animationController);
+ 
+    // Functionally equivalent to Android 12's SkSL:
+    //`return mix(u_touch, u_resolution, saturate(in_radius_scale * 2.0))`
+    final centerTween = Tween<Vector2>(
+      begin: Vector2.array([_position.dx, _position.dy]),
+      end: Vector2.array([referenceBox.size.width / 2, referenceBox.size.height / 2]),
+    );
+    final centerProgress = TweenSequence<double>(
+      <TweenSequenceItem<double>>[
+        TweenSequenceItem<double>(
+          tween: Tween<double>(begin: 0.0, end: 1.0),
+          weight: 50,
+        ),
+        TweenSequenceItem<double>(
+          tween: ConstantTween<double>(1.0),
+          weight: 50,
+        ),
+      ],
+    ).animate(_radiusScale);
+    _center = centerTween.animate(centerProgress);
+
+    _alpha = TweenSequence<double>(
+      <TweenSequenceItem<double>>[
+        TweenSequenceItem<double>(
+          tween: Tween<double>(begin: 0.0, end: 1.0),
+          weight: 13,
+        ),
+        TweenSequenceItem<double>(
+          tween: ConstantTween<double>(1.0),
+          weight: 27,
+        ),
+        TweenSequenceItem<double>(
+          tween: Tween<double>(begin: 1.0, end: 0.0),
+          weight: 60,
+        ),
+      ],
+    ).animate(_animationController);
+    
+    _sparkleAlpha = TweenSequence<double>(
+      <TweenSequenceItem<double>>[
+        TweenSequenceItem<double>(
+          tween: Tween<double>(begin: 0.0, end: 1.0),
+          weight: 13,
+        ),
+        TweenSequenceItem<double>(
+          tween: ConstantTween<double>(1.0),
+          weight: 27,
+        ),
+        TweenSequenceItem<double>(
+          tween: Tween<double>(begin: 1.0, end: 0.0),
+          weight: 50,
+        ),
+      ],
+    ).animate(_animationController);
+
+    // Creates an element of randomness so that ink eminating from the same
+    // pixel have slightly different rings and sparkles.
+    _turbulenceSeed = math.Random().nextDouble() * 1000.0;
```

Comment:
Oh, I see you intend to mock random below for the test. 👍

Reply:
Rather than mocking Random(), added a turbulence seed so that it can be constant, and a corresponding splashFactory. This is used for goldens, and can also be used if a developer wants a more deterministic effect.

I think mocking Random() would depend on mockito, but that doesn't appear to be a dep.